R/arx_forecaster_mod.R

```diff
@@ -1,29 +1,83 @@
-arx_epi_forecaster <- function(epi_data, response,
-                               ...,
+#' Direct autoregressive forecaster with covariates
+#'
+#' This is an autoregressive forecasting model for
+#' [epiprocess::epi_df] data. It does "direct" forecasting, meaning
+#' that it estimates a model for a particular target horizon.
+#'
+#'
+#' @param epi_data An `epi_df` object
+#' @param outcome A character (scalar) specifying the outcome (in the
+#'   `epi_df`).
+#' @param predictors A character vector giving column(s) of predictor
+#'   variables.
+#' @param trainer A `{parsnip}` model describing the type of estimation.
+#'   For now, we enforce `mode = "regression"`.
+#' @param args_list A list of customization arguments to determine
+#'   the type of forecasting model. See [arx_args_list()].
+#'
+#' @return A list with (1) `predictions` an `epi_df` of predicted values
+#'   and (2) `epi_workflow`, a list that encapsulates the entire estimation
+#'   workflow
+#' @export
+#'
+#' @examples
+#' jhu <- case_death_rate_subset %>%
+#'   dplyr::filter(time_value >= as.Date("2021-12-01"))
+#'
+#' out <- arx_epi_forecaster(jhu, "death_rate",
+#'   c("case_rate", "death_rate"))
+arx_epi_forecaster <- function(epi_data,
+                               outcome,
+                               predictors,
                                trainer = parsnip::linear_reg(),
                                args_list = arx_args_list()) {
 
-  r <- epi_recipe(epi_data) %>%
-    step_epi_lag(..., lag = args_list$lags) %>% # hmmm, same for all predictors
-    step_epi_ahead(response, ahead = args_list$ahead) %>%
-  # should use the internal function (in an open PR)
-    recipes::step_naomit(recipes::all_predictors()) %>%
-    recipes::step_naomit(recipes::all_outcomes(), skip = TRUE)
+  validate_forecaster_inputs(epi_data, outcome, predictors)
+  if (!is.list(trainer) || trainer$mode != "regression")
+    cli_stop("{trainer} must be a `parsnip` method of mode 'regression'.")
+  lags <- arx_lags_validator(predictors, args_list$lags)
+
+  r <- epi_recipe(epi_data)
+  for (l in seq_along(lags)) {
+    p <- predictors[l]
+    r <- step_epi_lag(r, !!p, lag = lags[[l]])
+  }
+  r <- r %>%
+    step_epi_ahead(dplyr::all_of(!!outcome), ahead = args_list$ahead) %>%
+    step_epi_naomit()
   # should limit the training window here (in an open PR)
   # What to do if insufficient training data? Add issue.
-  # remove intercept? not sure how this is implemented in tidymodels
+
+  forecast_date <- args_list$forecast_date %||% max(epi_data$time_value)
+  target_date <- args_list$target_date %||% forecast_date + args_list$ahead
   f <- frosting() %>%
     layer_predict() %>%
-    layer_naomit(.pred) %>%
-    layer_residual_quantile(
```

Comment:
Did this used to be a function or was this a typo before?

Reply:
I think that was the previous function name for `layer_residual_quantiles`, so it should've been updated to `layer_residual_quantiles`.

Reply:
Right, and I actually DON'T want to omit any NA's

Reply:
But leaving for historical reasons (and in case there's an example of where I'd want it).